### PR TITLE
Polishing and refactoring

### PR DESCRIPTION
**Pull request by segsell**
- [x] unindent computation of mariginal utilities
- [x] use dummy function for upper envelope if only one discrete choice
- [x] switch to three arrays for ```endog_grid```, ```policy```, and ```value``` instead of two big ones, i.e. only ```policy``` and ```value```, arrays with extra dimension for ```endog_grid```
- [x] harmonize outputs of ```final_period_wrapper```
- [x] update doscstrings

### CODE REVIEW

File: src/dcegm/solve.py [hunk 1]
```diff
@@ -124,32 +118,35 @@ def solve_dcegm(
...
+        transition_vector_by_state=transition_vector_by_state,
+        endog_grid_container=endog_grid_container,
+        policy_container=policy_container,
+        value_container=value_container,
         compute_upper_envelope=compute_upper_envelope,
         final_period_partial=final_period_partial,
     )
 
-    return policy_array, value_array
+    return endog_grid_container, policy_container, value_container
```
Comment (MaxBlesch) [inline src/dcegm/solve.py:149]: Adapt type hints of return.

File: src/dcegm/pre_processing.py [hunk 2]
```diff
@@ -105,12 +109,19 @@ def get_partial_functions(
...
+    else:
+        compute_upper_envelope = fast_upper_envelope_wrapper
+
     return (
         compute_utility,
         compute_marginal_utility,
         compute_inverse_marginal_utility,
         compute_value,
         compute_next_period_wealth,
+        compute_upper_envelope,
```
Comment (MaxBlesch) [inline src/dcegm/pre_processing.py:125]: Adapt type hints!

File: src/dcegm/egm.py [hunk 3]
```diff
@@ -44,34 +44,35 @@ def compute_optimal_policy_and_value(
...
-        compute_value,
-    )
+    policy = np.zeros(n_grid_wealth + 1)
+    policy[1:] = _policy
+
+    value = np.zeros(n_grid_wealth + 1)
+    value[0] = expected_value[0]
+    value[1:] = compute_value(_policy, expected_value, choice)
 
-    return current_policy_arr, current_value_arr
```
Comment (MaxBlesch) [inline src/dcegm/egm.py:74]: Add type hints!

**Review by MaxBlesch — CHANGES_REQUESTED** [review]
Correct type hints for changing to three arrays. Otherwise great work!

**Review by segsell — COMMENTED** [review]
Addressed comments

**Pull request merged**